inst/stan/functions/convolve.stan

```diff
@@ -1,29 +1,16 @@
-// convolve two vectors
-// length of x
-// produces a convolution across the lenght specified
-vector convolve(vector x, vector y, int len) {
-  int xlen = num_elements(x);
-  int ylen = num_elements(y);
-  vector[len] convolution = rep_vector(0, len);
-  for (i in 1:xlen) {
-    for (j in 1:(min(len - i + 1, ylen))) {
-      convolution[i + j - 1] += x[i] * y[j];
-    }
-  }
-  return(convolution);
-}
-
 // convolve two vectors as a backwards dot product
 // y vector shoud be reversed
 // limited to the length of x and backwards looking for x indexes
 // designed for use convolve a case vector and a delay pmf
-vector convolve_dot_product(vector x, vector y, int len) {
+vector convolve_with_rev_pmf(vector x, vector y, int len) {
+    int xlen = num_elements(x);
     int ylen = num_elements(y);
     vector[len] z;
     for (s in 1:len) {
-        z[s] = dot_product(
-            x[max(1, (s - ylen + 1)):s], tail(y, min(ylen, s))
-          );
+      z[s] = dot_product(
+        x[max(1, (s - ylen + 1)):min(s, xlen)],
+        y[max(1, ylen - s + 1):min(ylen, ylen + xlen - s)]
```

Comment:
does this slow down vs tail? Can tail not still be used here?

Reply:
Yeah I like the extra features. Trying to understand why it's the models without delays that suffer the big performance hits. Well small but non-trivial.

Reply:
[Profiling](https://github.com/epiforecasts/EpiNow2/blob/cmdstanr/inst/stan/estimate_infections.stan) suggests that everything involving convolutions is quite drastically slowed down.

PR:
```
                    name total_time
 1:                    R       4.30
 2:           assign max       0.00
 3:      day of the week       1.43
 4:               delays       3.63
 5:            delays lp       0.10
 6: generated quantities       0.01
 7:                gp lp       0.51
 8:                   gt       3.45
 9:           infections      10.31
10:            report lp       5.47
11:              reports      14.75
12:                rt lp       0.69
13:            update gp       4.47
```

main:
```
                    name total_time
 1:                    R       3.59
 2:     assign fixed pmf       0.00
 3:           assign max       0.00
 4:      day of the week       1.14
 5:               delays       0.59
 6:            delays lp       0.11
 7: generated quantities       0.01
 8:                gp lp       0.30
 9:                   gt       0.52
10:                gt lp       0.07
11:           infections       8.33
12:            report lp       4.58
13:              reports       6.01
14:                rt lp       0.53
15:             trunc lp       0.08
16:            update gp       3.36
```

Reply:
No delays model.

PR:
```
                    name total_time
 1:                    R       6.70
 2:           assign max       0.00
 3:      day of the week       2.37
 4:            delays lp       0.16
 5: generated quantities       0.01
 6:                gp lp       0.66
 7:                   gt       5.27
 8:           infections      16.54
 9:            report lp       8.84
10:              reports       0.24
11:                rt lp       1.08
12:            update gp       6.53
```

Main:
```
                    name total_time
 1:                    R       6.00
 2:     assign fixed pmf       0.00
 3:           assign max       0.00
 4:      day of the week       2.19
 5:               delays       0.19
 6:            delays lp       0.19
 7: generated quantities       0.01
 8:                gp lp       0.77
 9:                   gt       0.87
10:           infections      13.43
11:            report lp       7.90
12:              reports       0.47
13:                rt lp       0.73
14:             trunc lp       0.13
15:            update gp       6.18
```

Reply:
> But I can check if a reformulation yields performance benefits.

Doesn't look like it so will revert to previous version and try something else.

Reply:
Not ideal at all. Will take a close look on Friday if not before

Reply:
So the key targets are the delay and infections profiling block with some commonality between them (i.e. likely convolutions) leading to the slowdown. I wonder how if any of this is just allocation costs etc and the difference decreases as you run larger models (not ideal but better)

Reply:
It's a little bit faster with the latest changes - will profile again.

Reply:
Here's another comparison of time spent in different parts of the code (using cmdstan profiler) based on 100 chains. Not clear to me at all why e.g. more time seems would be spent calculating the reporting log-prior which hasn't been touched at all.

```
                    name mean.main mean.PR sd.main sd.PR
 1:                total     22.00   29.05    2.70  4.50
 2:            report lp      6.28    8.04    0.76  1.25
 3:           infections      6.07    8.83    0.75  1.36
 4:                    R      3.44    4.44    0.43  0.70
 5:            update gp      3.14    3.92    0.39  0.62
 6:      day of the week      1.11    1.41    0.14  0.21
 7:                   gt      0.55    1.41    0.07  0.22
 8:              reports      0.37      NA    0.05    NA
 9:                gp lp      0.37    0.46    0.05  0.08
10:                rt lp      0.33    0.47    0.04  0.07
11:            delays lp      0.17    0.08    0.03  0.01
12:               delays      0.10      NA    0.01    NA
13:                gt lp      0.05      NA    0.01    NA
14: generated quantities      0.01    0.01    0.00  0.00
15:     assign fixed pmf      0.00      NA    0.00    NA
16:           assign max      0.00    0.00    0.00  0.00
```